csharp/src/Ice/OutgoingFrame.cs

```diff
@@ -0,0 +1,169 @@
+        {
+            get
+            {
+                if (_payload == null && PayloadEnd is OutputStream.Position payloadEnd)
+                {
+                    var payload = new List<ArraySegment<byte>>();
+                    if (PayloadStart.Segment == payloadEnd.Segment)
+                    {
+                        payload.Add(Data[PayloadStart.Segment].Slice(PayloadStart.Offset,
+                                                                     payloadEnd.Offset - PayloadStart.Offset));
+                    }
+                    else
+                    {
+                        ArraySegment<byte> segment = Data[PayloadStart.Segment].Slice(PayloadStart.Offset);
+                        if (segment.Count > 0)
+                        {
+                            payload.Add(segment);
+                        }
+                        for (int i = PayloadStart.Segment + 1; i < payloadEnd.Segment; i++)
+                        {
+                            payload.Add(Data[i]);
+                        }
+
+                        segment = Data[payloadEnd.Segment].Slice(0, payloadEnd.Offset);
+                        if (segment.Count > 0)
+                        {
+                            payload.Add(segment);
+                        }
+                    }
+                    _payload = payload;
+                }
+                return _payload ?? Array.Empty<ArraySegment<byte>>();
+            }
+        }
+
+        /// <summary>The Ice protocol of this frame.</summary>
+        public Protocol Protocol { get; }
+
+        /// <summary>The frame byte count.</summary>
+        public int Size { get; protected set; }
+
+        // True if Ice1 frames should use protocol compression, false otherwise.
+        internal bool Compress { get; }
+        // Contents of the Frame
+        internal List<ArraySegment<byte>> Data { get; }
+        // Position of the end of the payload, for ice1 this is always the frame end.
+        internal OutputStream.Position? PayloadEnd;
+        // Position of the start of the payload.
+        internal OutputStream.Position PayloadStart;
+
+        private IList<ArraySegment<byte>>? _payload;
+
+        /// <summary>Compress the encapsulation payload using GZip compression, compressed encapsulation payload is
+        /// only supported with 2.0 encoding.</summary>
+        public void CompressPayload()
+        {
+            // TODO compress payload does nothing if Encoding != 2.0, should we throw instead?
+            if (Encoding == Encoding.V2_0)
+            {
+                if (PayloadEnd == null)
+                {
+                    throw new InvalidOperationException("payload has not been written");
+                }
+
+                IList<ArraySegment<byte>> payload = Payload;
+                int sizeLength = 1 << (payload[0][0] & 0x03);
+
+                byte compressionStatus = payload[0].Count > sizeLength + 2 ?
```

Comment:
Good catch, I missed that when fixing the reading/writing of the size.

The first segment is always 256 bytes, but we first write the identity/facet/operation/operationMode and context if Ice1, so it can still happen than the encaps header spawn several segments

Reply:
Fixed the computation of `sizeLenght` to check the Protocol Encoding